lib/spack/spack/version.py

```diff
@@ -603,7 +606,18 @@ def satisfies(self, other):
         self_cmp = self._cmp(other.ref_lookup)
         other_cmp = other._cmp(self.ref_lookup)
 
-        if other.is_ref:
+        if self.is_ref and other.is_ref:
+            if self.ref != other.ref:
+                return False
+            elif self.user_supplied_reference and other.user_supplied_reference:
+                return self.ref_version == other.ref_version
+            elif other.user_supplied_reference:
+                return False
+            else:
+                # In this case, neither supplies a version equivalence with "="
+                # and the commit strings are equal
+                return True
+        elif other.is_ref:
```

Comment:
@alalazo @tgamblin I think this implements the desired semantics and I added `test_git_user_supplied_reference_satisfaction` to confirm that (I didn't see a test covering these in `test/versions.py`). I don't think this will conflict with  https://github.com/spack/spack/pull/35681 (that doesn't change `satisfies`, and  furthermore `satisfies` can be more strict than `intersects`).